File: packages/modules/clipboard/ios/ClipboardModule.swift
import StingRuntime
import UIKit

public final class ClipboardModule: StingNativeModule {
    public let name = "Clipboard"
    public let version = "0.1.0"

    public init() {}

    public func callSync(method: String, arguments: [Any]) throws -> Any? {
        switch method {
        case "getString":
            return UIPasteboard.general.string ?? ""
        case "setString":
            guard let value = arguments.first as? String else {
                throw StingNativeModuleError(
                    code: "E_INVALID_ARGUMENT",
                    message: "Clipboard.setString requires a string"
                )
            }
            UIPasteboard.general.string = value
            return nil
        case "hasString":
            return !(UIPasteboard.general.string ?? "").isEmpty
        case "clear":
            UIPasteboard.general.items = []
            return nil
        default:
            throw StingNativeModuleError(
                code: "E_METHOD_NOT_FOUND",
                message: "Clipboard does not implement \(method)"
            )
        }
    }
}

File: packages/modules/clipboard/android/src/main/java/run/stingjs/modules/clipboard/ClipboardModule.kt
package run.stingjs.modules.clipboard

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import run.stingjs.runtime.StingNativeModule
import run.stingjs.runtime.StingNativeModuleError

class ClipboardModule(context: Context) : StingNativeModule {
    override val name = "Clipboard"
    override val version = "0.1.0"

    private val appContext = context.applicationContext
    private val clipboard = appContext.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

    override fun callSync(method: String, arguments: List<Any?>): Any? = when (method) {
        "getString" -> readString()
        "setString" -> {
            val value = arguments.firstOrNull() as? String
                ?: throw StingNativeModuleError(
                    code = "E_INVALID_ARGUMENT",
                    message = "Clipboard.setString requires a string",
                )
            clipboard.setPrimaryClip(ClipData.newPlainText("StingJS", value))
            null
        }
        "hasString" -> readString().isNotEmpty()
        "clear" -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                clipboard.clearPrimaryClip()
            } else {
                clipboard.setPrimaryClip(ClipData.newPlainText("", ""))
            }
            null
        }
        else -> throw StingNativeModuleError(
            code = "E_METHOD_NOT_FOUND",
            message = "Clipboard does not implement $method",
        )
    }

    private fun readString(): String {
        val clip = clipboard.primaryClip ?: return ""
        if (clip.itemCount == 0) return ""
        return clip.getItemAt(0).coerceToText(appContext)?.toString().orEmpty()
    }
}
